docs/layout-of-source-files.rst

```diff
@@ -216,101 +216,562 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
-it is not considered as the current or the parent directory.
-All path names are treated as absolute paths unless they start with the current ``.`` or the parent directory ``..``.
+Path Resolution
+---------------
 
-To import a file ``filename`` from the same directory as the current file, use ``import "./filename" as symbolName;``.
-If you use ``import "filename" as symbolName;`` instead, a different file could be referenced
-(in a global "include directory").
+The paths used in imports are not filesystem paths.
+In the simplest cases they may end up being used directly to load files from disk but in general
+the final path can be substantially different due to abstractions necessary to ensure reproducible
+builds on different platforms.
 
-It depends on the compiler (see :ref:`import-compiler`) how to actually resolve the paths.
-In general, the directory hierarchy does not need to strictly map onto your local
-filesystem, and the path can also map to resources such as ipfs, http or git.
+.. index:: import key, import path, filesystem path
 
-.. note::
-    Always use relative imports like ``import "./filename.sol";`` and avoid
-    using ``..`` in path specifiers. In the latter case, it is probably better to use
-    global paths and set up remappings as explained below.
+Virtual Filesystem
+~~~~~~~~~~~~~~~~~~
+
+The compiler maintains an internal database (virtual filesystem) where each compiled module is
+assigned a unique *import key* which is an opaque and unstructured identifier.
+
+While an import key in the virtual filesystem can be anything, it should be a valid path if the
+source is meant to be loaded from the underlying filesystem.
+When the requested key is not present in the virtual filesystem, it is passed to the file loader.
+In case of the command-line compiler the file loader simply uses it as a path.
+The `JavaScript interface <https://github.com/ethereum/solc-js>`_ is a bit more flexible in that
+regard and allows the user to provide a callback to perform this operation.
+In this case the keys can be arbitrary.
+For example they could be URLs as long as the custom loader can handle them.
+
+There are several ways to load modules into the virtual filesystem:
+
+#. **import statement**
+
+   The ``import`` statement tells the compiler to load and compile a module and then make its
+   content available to the current module.
+
+   We will refer to the path used in the statement as *import path*.
+   The import path is translated into an import key first and then the compiler uses the key to
+   look up the file in its virtual filesystem.
+
+   Imports can be broadly classified into three categories based on how the path is specified:
+
+   .. code-block:: solidity
+
+       import "/contracts/lib/token.sol";   // Absolute
+       import "contracts/lib/token.sol";    // Relative to base
+       import "./contracts/lib/token.sol";  // Relative to source
+
```

Comment:
What about `import "../a/b.sol";`?

Reply:
I could add it but there's a whole section about relative imports below with many examples so I tried to keep it short there.

Reply:
Done.